### PR TITLE
Content Manager - Add usage documentation

### PR DESCRIPTION
**Pull request by tsarquis88**
|Related issue|
|---|
|#20160 |

## Description

This PR adds a README.md file to the Content Manager module with a `Usage` section. This section helps users to understand how to configure and use the Content Manager module.

### CODE REVIEW

File: src/shared_modules/content_manager/README.md [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+The Content Manager counts with a [test tool](./testtool/main.cpp) that can be used to perform tests, try out different configurations, and to better understand the module.
+
+The input configuration of the Content Manager is described below:
+
+- `topicName`: Topic name, used to represent the actions being executed.
+- `interval`: Interval, in seconds, between each action execution.
+- `ondemand`: If true, the module will be executed on demand.
+- `configData`: Configuration data to create the orchestration of the module.
+  + `contentSource`: Source of the content. Can be any of `api`, `cti-api`, `file`, or `offline`.
+  + `compressionType`: Compression type of the content. Can be any of `gzip`, `zip`, or `xz`.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:16]: Improve: It can also be `raw`
```suggestion
  + `compressionType`: Compression type of the content. Can be any of `gzip`, `zip`, `xz` or `raw`.
```

File: src/shared_modules/content_manager/README.md [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+  + `versionedContent`: Type of versioned content. Can be any of `false` (content versioning disabled) or `cti-api`.
+  + `deleteDownloadedContent`: If true, the downloaded content will be deleted after being processed.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:18]: Nitpick (clarity): Highlight possible values
```suggestion
  + `deleteDownloadedContent`: If `true`, the downloaded content will be deleted after being processed.
```

File: src/shared_modules/content_manager/README.md [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+  + `url`: URL from where the content will be downloaded or copied.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:19]: Nitpick: Maybe explaining that it supports http, https or file?

File: src/shared_modules/content_manager/README.md [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+  + `outputFolder`: If defined, the content (downloads and uncompressed content) will be downloaded in this folder.
+  + `dataFormat`: Content data format. Examples: `json`, `xml`, `txt`, etc.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:21]: Improve: How many other formats are supported? I would remove the etc all together as it adds no useful information

File: src/shared_modules/content_manager/README.md [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+  + `contentFileName`: Used by some downloaders to know where to store the downloaded content.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:22]: I didn't understand this one, its the name of the downloaded content?

File: src/shared_modules/content_manager/README.md [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+  "userId": 1,
+  "id": 1,
+  "title": "delectus aut autem",
+  "completed": false
+}
+```
+
+> If the content was compressed, the output file would be stored in the _downloads_ folder.
+
+The `/tmp/output_folder/contents/content.json` path is published for the consumers to read.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:193]: Nitpick: I think subscribe aligns better with the underlying architecture of the module
```suggestion
The `/tmp/output_folder/contents/content.json` path is published for the consumers to subscribe.
```

File: src/shared_modules/content_manager/README.md [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+### Use case: Remote file download
+
+The Content Manager has the capability of downloading a content file from a URL. The file will be stored in the output folder and, if compressed, it will be decompressed in the contents folder.
```
Comment (sebasfalcone) [inline src/shared_modules/content_manager/README.md:197]: Improve (clarity): I think this is a bit hard to follow, maybe show the tree structure along side (also applies to the other examples)

**Review by sebasfalcone — COMMENTED** [review]
Partial review

File: src/shared_modules/content_manager/README.md [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+CtiApiDownloader - Request processed successfully.
+CtiApiDownloader - Request processed successfully.
+CtiApiDownloader - Request processed successfully.
+CtiApiDownloader - Request processed successfully.
+CtiApiDownloader - Finishing
+SkipStep - Executing
+PubSubPublisher - Data published
+```
+
+The output files containing the downloaded offsets are available under `output_folder/contents/`, each of which contains 1000 offsets (the last one can contain fewer offsets if the total amount of offsets is not multiple of 1000).
```
Comment (jftuduri) [inline src/shared_modules/content_manager/README.md:80]: Question: The 1000 is hardcoded in the content manager? or is it a limitation of the API?

**Review by kevin-mcm — COMMENTED** [review]
shouldn't this file be inside the `content_manager/doc` directory?

File: src/shared_modules/content_manager/README.md (hunk 4 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:21]: In fact, any `dataFormat` is supported, that's why the "etc.". The `dataFormat` is used to make a replacement in some decompressors for the output files.

So, leaving just "json, xml, txt" doesn't seem right. I could remove all of them, as an alternative. 
What do you think?

File: src/shared_modules/content_manager/README.md (hunk 5 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:22]: Yes, this variable is used to form the output filename. I will rephrase it.

File: src/shared_modules/content_manager/README.md (hunk 6 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:193]: Consumers don't subscribe themselves to a path but to a channel.

File: src/shared_modules/content_manager/README.md (hunk 7 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:197]: I'm already providing the tree structure of the output in every example. You have to scroll down a little bit though.

File: src/shared_modules/content_manager/README.md [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+
+The input configuration of the Content Manager is described below:
+
+- `topicName`: Topic name, used to represent the actions being executed.
+- `interval`: Interval, in seconds, between each action execution.
+- `ondemand`: If true, the module will be executed on demand.
+- `configData`: Configuration data to create the orchestration of the module.
+  + `contentSource`: Source of the content. Can be any of `api`, `cti-api`, `file`, or `offline`.
+  + `compressionType`: Compression type of the content. Can be any of `gzip`, `zip`, or `xz`.
+  + `versionedContent`: Type of versioned content. Can be any of `false` (content versioning disabled) or `cti-api`.
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:17]: Yes... and no :sweat_smile:

If we enable the `cti-api` content versioner with a downloader different from `cti-api`, the version updater will be instantiated and will update the database offset (offset that shouldn't have changed since there is no `cti-api` downloader).

I will add a comment to clarify this.

File: src/shared_modules/content_manager/README.md (hunk 8 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:80]: Both. It is hardcoded because of an API limitation. Do you think it's worth to mention this?

The limitation is given by the Wazuh CTI API I believe. Maybe (we never know) a user wants to use its own CTI API, so a comment saying that "there is a limitation from the API" doesn't seem right to me.

**Comment by tsarquis88** [comment]
> shouldn't this file be inside the `content_manager/doc` directory?

@kevin-mcm Maybe.

I thought the Content Manager as a whole project and, as such, it should have a "main" README file IMO (like the [Wazuh readme](https://github.com/wazuh/wazuh/blob/master/README.md)).

File: src/shared_modules/content_manager/README.md (hunk 3 above)
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: Done.

File: src/shared_modules/content_manager/README.md [hunk 10]
```diff
@@ -0,0 +1,341 @@
+# Content Manager
+
+The Content Manager is a module that is in charge of obtaining the Wazuh Content from an external repository, as well as maintaining the periodic updates that the manager will obtain.
+
+## Usage
+
+The Content Manager counts with a [test tool](./testtool/main.cpp) that can be used to perform tests, try out different configurations, and to better understand the module.
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: Done.

File: src/shared_modules/content_manager/README.md [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+
+The input configuration of the Content Manager is described below:
+
+- `topicName`: Topic name, used to represent the actions being executed.
+- `interval`: Interval, in seconds, between each action execution.
+- `ondemand`: If true, the module will be executed on demand.
+- `configData`: Configuration data to create the orchestration of the module.
+  + `contentSource`: Source of the content. Can be any of `api`, `cti-api`, `file`, or `offline`.
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: Done.

File: src/shared_modules/content_manager/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: Done.

File: src/shared_modules/content_manager/README.md [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+  + `compressionType`: Compression type of the content. Can be any of `gzip`, `zip`, or `xz`.
+  + `versionedContent`: Type of versioned content. Can be any of `false` (content versioning disabled) or `cti-api`.
+  + `deleteDownloadedContent`: If true, the downloaded content will be deleted after being processed.
+  + `url`: URL from where the content will be downloaded or copied.
+  + `outputFolder`: If defined, the content (downloads and uncompressed content) will be downloaded in this folder.
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md:18]: Good question. Added a comment about the default path.

File: src/shared_modules/content_manager/README.md [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+  + `dataFormat`: Content data format. Examples: `json`, `xml`, `txt`, etc.
+  + `contentFileName`: Used by some downloaders to know where to store the downloaded content.
+  + `databasePath`: Path from where the RocksDB database should be read. The database stores the last offset fetched (when using the `cti-api` content source).
+
+### Use case: Download offsets from CTI API
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: Done!

File: src/shared_modules/content_manager/README.md [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+- `configData`: Configuration data to create the orchestration of the module.
+  + `contentSource`: Source of the content. Can be any of `api`, `cti-api`, `file`, or `offline`.
+  + `compressionType`: Compression type of the content. Can be any of `gzip`, `zip`, or `xz`.
+  + `versionedContent`: Type of versioned content. Can be any of `false` (content versioning disabled) or `cti-api`.
+  + `deleteDownloadedContent`: If true, the downloaded content will be deleted after being processed.
+  + `url`: URL from where the content will be downloaded or copied.
+  + `outputFolder`: If defined, the content (downloads and uncompressed content) will be downloaded in this folder.
+  + `dataFormat`: Content data format. Examples: `json`, `xml`, `txt`, etc.
+  + `contentFileName`: Used by some downloaders to know where to store the downloaded content.
+  + `databasePath`: Path from where the RocksDB database should be read. The database stores the last offset fetched (when using the `cti-api` content source).
```
[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: True, done.

File: src/shared_modules/content_manager/README.md (hunk 8 above)
Comment (jftuduri) [reply src/shared_modules/content_manager/README.md:89]: Thanks for the answer.
I think we can leave it as it is.

File: src/shared_modules/content_manager/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (jftuduri) [reply src/shared_modules/content_manager/README.md]: > If we enable the cti-api content versioner with a downloader different from cti-api, the version updater will be instantiated and will update the database offset (offset that shouldn't have changed since there is no cti-api downloader).

But is that a problem? I mean, setting a non-cti-api downloader and setting versionedContent with cti-api: is it a config error? should it be a config error? will something break?

[the remark this answers is not in the document]
Comment (tsarquis88) [reply src/shared_modules/content_manager/README.md]: It's not a config error, nor will end up breaking anything.
It will just execute a version updater that won't have any purpose (updating an offset that didn't change).

Should it be a config error? Not that I know of. Maybe @Damian-Mangold can tell.

**Review by jftuduri — APPROVED** [review]
LGTM

**Review by kevin-mcm — APPROVED** [review]

**Review by sebasfalcone — APPROVED** [review]
LGTM

**Comment by tsarquis88** [comment]
Please @wazuh/drop-table, review the last commit.

**Review by jftuduri — APPROVED** [review]

**Review by sebasfalcone — APPROVED** [review]
LGTM

**Review by kevin-mcm — APPROVED** [review]

**Review by Damian-Mangold — APPROVED** [review]
Excellent !!
